include/swift/AST/Types.h

```diff
@@ -4308,7 +4308,8 @@ class SILFunctionType final : public TypeBase, public llvm::FoldingSetNode,
       IndexSubset *parameterIndices, unsigned resultIndex,
       AutoDiffDerivativeFunctionKind kind, Lowering::TypeConverter &TC,
       LookupConformanceFn lookupConformance,
-      CanGenericSignature derivativeFunctionGenericSignature = nullptr);
+      CanGenericSignature derivativeFunctionGenericSignature = nullptr,
+      bool isReabstractionThunk = false);
```

Comment:
> It would be helpful to have a documentation comment explaining what the derivative function of a reabstraction thunk is.

Added doc comment for `SILFunctionType::getAutoDiffDerivativeFunctionType` in https://github.com/apple/swift/pull/28570/commits/6eb7a070fd6734c1deb1220556038e4267d349ef:
```cpp
  /// Special cases:
  /// - Reabstraction thunks have special derivative type calculation. The
  ///   original function-typed last parameter is transformed into a
  ///   `@differentiable` function-typed parameter in the derivative type. This
  ///   is necessary for the differentiation transform to support reabstraction
  ///   thunk differentiation because the function argument is opaque and cannot
  ///   be differentiated. Instead, the argument is made `@differentiable` and
  ///   reabstraction thunk JVP/VJP callers are reponsible for passing a
  ///   `@differentiable` function.
```

Also added comment in `reapplyFunctionConversion` in Differentiation.cpp explaining `partial_apply` reapplication for reabstraction thunk derivatives.

> If I understand correctly, it is "a function that, when partially applied to the original function, becomes the derivative function of the reabstracted function." Or maybe instead it "becomes a reabstraction of the derivative function". I haven't thought carefully enough to figure out which one it is, or if there is a difference, but maybe there are situations where the difference is important.

I believe your understanding is correct. I made a [short write-up](https://docs.google.com/document/d/1_BirmTqdotglwNTOcYAW-ib6mx_jl-gH9Dbg4WmHZh0/edit#bookmark=id.inyyvy1wapx) about the reabstraction thunk differentiation approach.

> Another question I have is: Why does the derivative function of the reabstraction thunk take the original function as argument rather than than the derivative function?

That's a good point! A reabstraction thunk derivative only needs the original function's corresponding derivative, not any other `@differentiable` function components. I think that approach is more efficient and semantically clear, so I'll pursue it now.

Reply:
Actually, to focus on retrodiff as a priority, how about deferring optimizations and using the current approach for now? TF-1036 tracks reabstraction thunk derivative generation optimizations.

Reply:
sounds good

nice doc comment!